### PR TITLE
Add Metrics Section to Datasource Id Page

### PR DESCRIPTION
**Pull request by mknowlton89**
### Features and Changes

This is a simple PR that adds a "Metrics" section to a datasource's id page (`/datasource/[did]`). 

To do this, I've set up a new endpoint that takes in a datasource's id and returns the metrics associated with that datasource.

We're then rendering the metrics on the page. I have matched the functionality within the `/metrics` page, so a user can click on a metric to go to it's metric id page, or via the `<MoreMenu />` a user can duplicate or archive a metric.

Additionally, I've add an "Add" button in the Metric section that utilizes the `MetricForm` so a user can add a new metric directly from the datasource id page, and we're locking in the `datasource` within the input field. 

In a future PR, this is where I'll introduce another entry point for auto generating metrics if the datasource supports it.

### Testing

- [x] Ensure that all metrics are rendered correctly when viewing a datasource id page
- [x] Ensure that if a datasource does not have any metrics associated with it, we're showing the correct empty state
- [x] Ensure that you can create a new metric directly from the datasource id page
- [x] Ensure that if the org is using the config file, or if the user doesn't have `createMetrics` permission, we're not showing the `Add` CTA
- [x] Ensure that if a metric is archived, the font is grayed out, and within the `<MoreMenu />` we're showing the "Unarchive" option
- [x] Ensure the duplicate option works as expected

### Screenshots

Screen Shot 2023-06-29 at 4 17 11 PM

Screen Shot 2023-06-29 at 3 51 19 PM

(`usingFileConfig()` === true`) - No "Add" CTA
Screen Shot 2023-06-29 at 4 17 29 PM

### CODE REVIEW

**Comment by lukesonnet** [comment]
Having this near the bottom makes sense, but I still think it's taking up too much real estate. Minor suggestion to do some combo of the following:
- Reduce info to one line, or a tabular view that is more compact, or have a compact/expanded view w/ toggle that we can use here and for other parts of the app (e.g. the metrics page itself). The latter is the most work but I think it'll also help with future changes where we want to improve experiment, metric, and feature discoverability
- Hide this in a collapsable element, maybe with text saying (3 existing) or something like that to summarize the N of metrics that you will see more details of when you click the element (or an arrow). Then if there are no metrics, you can hijack that element title/text to call this out & have the CTA for auto-metrics and/or adding metrics.

**Comment by mknowlton89** [comment]
> Having this near the bottom makes sense, but I still think it's taking up too much real estate. Minor suggestion to do some combo of the following:
> 
> * Reduce info to one line, or a tabular view that is more compact, or have a compact/expanded view w/ toggle that we can use here and for other parts of the app (e.g. the metrics page itself). The latter is the most work but I think it'll also help with future changes where we want to improve experiment, metric, and feature discoverability
> * Hide this in a collapsable element, maybe with text saying (3 existing) or something like that to summarize the N of metrics that you will see more details of when you click the element (or an arrow). Then if there are no metrics, you can hijack that element title/text to call this out & have the CTA for auto-metrics and/or adding metrics.

@lukesonnet Thanks for your feedback!

I worked through some of your suggestions this morning to come up with a new treatment for these sections. Let me know what you think.

A few caveats
- I added the `Discover Metrics` button as that will soon live on this page, so I wanted to mock it up with that in place, but that work will be done in a separate PR
- Depending on the direction we go, we'll likely want to update the treatment for the other sections (`Identifiers`, `Assignment Queries`, and `Jupyter Notebooks`) so everything is uniform. _That's probably understood, but just wanted to call it out_
- I looked into adding a tooltip next to the `Metrics` headline, but it was competing with the pill for the number of metrics to reduce the amount of text on the screen. 

### Screenshots
Closed, with no metrics
Screen Shot 2023-06-30 at 9 27 30 AM

Open, with no metrics
Screen Shot 2023-06-30 at 9 27 38 AM

Open, with metrics
Screen Shot 2023-06-30 at 9 27 54 AM

**Comment by lukesonnet** [comment]
The one tiny bit of design feedback about getting things to be more uniform would be making the other sections in little cards as well, even if we don't add drop downs for them. Right now Metrics is a whole section elevated to that card when the others are just headers.

**Comment by mknowlton89** [comment]
> This looks great! I think `Metrics` might be treated separately from the others because some users will have 1000 metrics on one datasource, but likely only a few experiment assignment queries and identifiers.

Sweet! Thanks again for taking a look yesterday and for suggesting the update.

I'll go back and revise the other sections so they're all more uniform and then let you know when it's ready for another peak.

Thanks!

**Comment by mknowlton89** [comment]
@lukesonnet Just pushed up some changes to make the datasource id page more uniform - curious to hear your thoughts on the updated styling - I did a few things.

1. The "Identifiers" headline was kinda redundant with the "Identifier Types" subhead, so I consolidated, and then broke out the "Joins" section to it's own block.

Before
Screen Shot 2023-06-30 at 4 01 17 PM

After
Screen Shot 2023-06-30 at 4 01 01 PM


2. I added `bg-white` to each section to try and visually differentiate each section more. One side effect of that was I had to make the individual cards within each section `bg-light`. I'm curious if you feel that makes them look as if they're somehow disabled.

Before
Screen Shot 2023-06-30 at 4 02 46 PM

After
Screen Shot 2023-06-30 at 4 03 02 PM

File: packages/front-end/pages/datasources/[did].tsx [hunk 1]
```diff
@@ -282,40 +331,239 @@ mixpanel.init('YOUR PROJECT TOKEN', {
...
+              <div className="my-3 p-3 rounded border bg-white">
                 <ExperimentAssignmentQueries
                   dataSource={d}
                   onSave={updateDataSourceSettings}
                   onCancel={() => undefined}
                   canEdit={canEdit}
                 />
               </div>
+              <div className="my-3 p-3 rounded border bg-white">
+                <div className="d-flex flex-row align-items-center justify-content-between">
```
Comment (jdorn) [inline packages/front-end/pages/datasources/[did].tsx:361]: Move this metrics section into its own self-contained component

File: packages/front-end/pages/datasources/[did].tsx [hunk 2]
```diff
@@ -37,19 +49,25 @@ function quotePropertyName(name: string) {
...
   const router = useRouter();
+  const [modalData, setModalData] = useState<{
+    current: Partial<MetricInterface>;
+    edit: boolean;
+    duplicate: boolean;
+  } | null>(null);
 
   const {
     getDatasourceById,
+    project,
```
[the remark this answers is not in the document]
Comment (mknowlton89) [reply packages/front-end/pages/datasources/[did].tsx:64]: @jdorn I'm not sure I completely grok the expected behavior here with this updated logic.

Before this, we were taking the currently selected project, and using that to determine whether the user can edit a metric, and whether or not they could create a metric.

How do we want to handle the add/edit CTA in the event a user only has permission to do so in 1 project, but the metric lives in 2 projects? It seems like we'd want to block them from being able to edit the metric, to not introduce any permission leaks.

File: packages/front-end/components/Metrics/MetricForm/index.tsx [hunk 3]
```diff
@@ -253,7 +253,11 @@ const MetricForm: FC<MetricFormProps> = ({
       timestampColumn: current.timestampColumn || "",
       tags: current.tags || [],
       projects:
-        edit || duplicate ? current.projects || [] : project ? [project] : [],
+        current.projects || edit || duplicate
```
Comment (bttf) [inline packages/front-end/components/Metrics/MetricForm/index.tsx:256]: Why was this change necessary? I'm wondering if it's because we have a new context to view this in that's not an 'edit' or 'duplicate' situation. Would checking `source` instead be a possible solution that would be more readable?
```suggestion
        source === 'datasource-detail' || edit || duplicate
```

File: packages/front-end/components/Settings/EditDataSource/DataSourceMetrics.tsx [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+                                  {metric.type}
+                                </code>
+                              </div>
+                              <div
+                                className={clsx(
+                                  metric.status === "archived"
+                                    ? "text-muted"
+                                    : "",
+                                  "pr-3"
+                                )}
```
Comment (bttf) [inline packages/front-end/components/Settings/EditDataSource/DataSourceMetrics.tsx:161]: Just a tip on how clsx can accomplish this more concisely, looks like there are a few opportunities to do this in this file

```suggestion
                                className={clsx({
                                    "text-muted": metric.status === "archived"
                                  },
                                  "pr-3"
                                )}
```

**Review by bttf — APPROVED** [review]
Some minor comments but looks good

File: packages/front-end/components/Metrics/MetricForm/index.tsx (hunk 3 above)
Comment (mknowlton89) [reply packages/front-end/components/Metrics/MetricForm/index.tsx:256]: @bttf This change adds logic so if we pass in `current.projects` to the form, we automatically seed the MetricForm with those projects. It was requested that when creates a new metric from the datasourceId page, that metric inherits the datasource's projects.

Checking for `source === 'datasource-detail'` does work here, in lieu of checking for `current.projects`. Updating.

Comment (mknowlton89) [reply packages/front-end/components/Metrics/MetricForm/index.tsx]: Addressed in this [commit](https://github.com/growthbook/growthbook/pull/1412/commits/f1affc08c34294aef3062e7f168a71af13fd0cce).

File: packages/front-end/components/Settings/EditDataSource/DataSourceMetrics.tsx (hunk 4 above)
Comment (mknowlton89) [reply packages/front-end/components/Settings/EditDataSource/DataSourceMetrics.tsx]: Addressed in this [commit](https://github.com/growthbook/growthbook/pull/1412/commits/f1affc08c34294aef3062e7f168a71af13fd0cce).
